crates/proto/src/event.rs

```diff
@@ -44,18 +39,18 @@ pub trait ProtoEvent: Message + Name + Sized {
             return Err(anyhow::anyhow!(format!("ABCI Event {} not expected for {}", event_type, Self::NAME)))
         }
 
-        let mut attributes_kv = HashMap::<String, Vec<u8>>::new();
+        // NOTE: Is there any condition where there would be duplicate EventAttributes and problems that fall out of that?
+        let attributes = event
+            .clone()
+            .attributes
+            .into_iter()
+            .map(|x| (x.key, serde_json::to_value(x.value).expect("EventAttribute Value should be Serializeable.")))
+            .collect::<HashMap<String, serde_json::Value>>();
 
-        for attr in &event.attributes {
-            let _ = attributes_kv.insert(
-                attr.key.clone(),
-                serde_json::to_vec(&attr.value).expect("EventAttribute values should be serializeable.")
-            );
-        }
+        let json = serde_json::to_value(attributes).expect("HashMap of String, serde_json::Value should be serializeable.");
 
-        // NOTE: Same thing here as into_event(), check what the equivalent to the jsonpb Unmarshaler is.
-        let json_bytes = serde_json::to_vec(&attributes_kv).expect("Hashmap of EventAttribute Keys and values should be serializeable.");
-        // let decoded_event = Self::decode(serde_json::from_slice(&json_bytes).expect("Serde derived serialized bytes should be deserializeable."));
+        // Hitting a lifetime error that requires further restriction on Self: Deserialize that conflicts with the current supertraits defined above.
+        // return Ok(serde_json::from_value(json)?)
```

Comment:
Ok, @hdevalence, all I'm left is figuring out how to get the final bit of equivalent unmarshalling/marshalling code in `from_event` method so I can do the actual test data/string munging of results. These lines are the rough equivalent to the final lines of the [go impl here](https://github.com/cosmos/cosmos-sdk/blob/8fb62054c59e580c0ae0c898751f8dc46044499a/types/events.go#L145-L155).

With the final bit twiddling, do I just want `serde_json` to give me back a value of `Result<Self>` like I have commented out above? If so, I'm hitting a Trait restriction compile error directly on `Self` that required an additional lifetime that had me re-writing the Trait's signature to the point I started leaking out the parameterized type variable into the `StateWriterTrait` trait.

I tried massaging things around but I want to make sure I'm not going on another snipe hunt again.

Reply:
What about using https://docs.rs/serde/latest/serde/de/trait.DeserializeOwned.html ?